src/io_darwin.zig

```diff
@@ -0,0 +1,691 @@
+            // - run_for_ns() always submits a timeout
+            if (change_events == 0 and self.completed.peek() == null) {
+                if (!wait_for_completions) return;
+                const timeout_ns = next_timeout orelse @panic("kevent() blocking forever");
+                ts.tv_nsec = @intCast(@TypeOf(ts.tv_nsec), timeout_ns % std.time.ns_per_s);
+                ts.tv_sec = @intCast(@TypeOf(ts.tv_sec), timeout_ns / std.time.ns_per_s);
+            }
+
+            const new_events = try os.kevent(
+                self.kq, 
+                events[0..change_events],
+                events[0..events.len],
+                &ts,
+            );
+
+            // Mark the io events submitted only after kevent() successfully processed them
+            self.io_pending.out = io_pending;
+            if (io_pending == null) {
+                self.io_pending.in = null;
+            }
+
+            for (events[0..new_events]) |event| {
+                const completion = @intToPtr(*Completion, event.udata);
+                self.completed.push(completion);
+            }
+        }
+
+        var completed = self.completed;
+        self.completed = .{};
+        while (completed.pop()) |completion| {
+            (completion.callback)(self, completion);
+        }
+    }
+
+    fn flush_io(self: *IO, events: []os.Kevent, io_pending_top: *?*Completion) usize {
+        for (events) |*event, flushed| {
+            const completion = io_pending_top.* orelse return flushed;
+            io_pending_top.* = completion.next;
+
+            const event_info = switch (completion.operation) {
+                .accept => |op| [2]c_int{ op.socket, os.EVFILT_READ },
+                .connect => |op| [2]c_int{ op.socket, os.EVFILT_WRITE },
+                .read => |op| [2]c_int{ op.fd, os.EVFILT_READ },
+                .write => |op| [2]c_int{ op.fd, os.EVFILT_WRITE },
+                .recv => |op| [2]c_int{ op.socket, os.EVFILT_READ },
+                .send => |op| [2]c_int{ op.socket, os.EVFILT_WRITE },
+                else => @panic("invalid completion operation queued for io"),
+            };
+
+            event.* = .{
+                .ident = @intCast(u32, event_info[0]),
+                .filter = @intCast(i16, event_info[1]),
+                .flags = os.EV_ADD | os.EV_ENABLE | os.EV_ONESHOT,
+                .fflags = 0,
+                .data = 0,
+                .udata = @ptrToInt(completion),
+            };
+        }
+        return events.len;
+    }
+    
+    fn flush_timeouts(self: *IO) ?u64 {
+        var min_timeout: ?u64 = null;
+        var timeouts: ?*Completion = self.timeouts.peek();
+        while (timeouts) |completion| {
+            timeouts = completion.next;
+
+            // NOTE: We could cache `now` above but monotonic() should be cheap to call ideally.
+            const now = self.time.monotonic();
+            const expires = completion.operation.timeout.expires;
```

Comment:
Nice to be able to use these monotonic timestamps for this. It's also cleaner keeping these timeouts out of the kernel entirely compared to our Linux IO where we are more coupled and need to work together with io_uring in using the same monotonic clock source.